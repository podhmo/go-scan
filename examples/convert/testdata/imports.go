package testdata

import (
	"time"

	"github.com/podhmo/go-scan/examples/convert/testdata/external"
)

// convert:import ext "github.com/podhmo/go-scan/examples/convert/testdata/external"
// convert:rule "time.Time" -> "string", using=ext.TimeToString
// convert:rule "string", validator=ext.ValidateString

// @derivingconvert(DstWithImports)
type SrcWithImports struct {
	Timestamp time.Time
	Comment   string
}

type DstWithImports struct {
	Timestamp string
	Comment   string
}
