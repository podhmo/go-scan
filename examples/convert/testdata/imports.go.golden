// Code generated by convert. DO NOT EDIT.
package main

import (
	"context"
	"errors"

	"github.com/podhmo/go-scan/examples/convert/model"
)

func convertSrcToDst(ctx context.Context, ec *model.ErrorCollector, src *Src) *Dst {
	if src == nil {
		return nil
	}
	dst := &Dst{}
	if ec.MaxErrorsReached() {
		return dst
	}
	ec.Enter("Timestamp")
	dst.Timestamp = src.Timestampext.ValidateString(ec, dst.Timestamp)
	ec.Leave()
	if ec.MaxErrorsReached() {
		return dst
	}
	ec.Enter("Comment")
	dst.Comment = src.Commentext.ValidateString(ec, dst.Comment)
	ec.Leave()
	return dst
}

// ConvertSrcToDst converts Src to Dst.
func ConvertSrcToDst(ctx context.Context, src *Src) (*Dst, error) {
	if src == nil {
		return nil, nil
	}
	ec := model.NewErrorCollector(0)
	dst := convertSrcToDst(ctx, ec, src)
	if ec.HasErrors() {
		return dst, errors.Join(ec.Errors()...)
	}
	return dst, nil
}
